File: src/main/java/ms/konovalov/intellij/hidpi/FontSizeComponent.java
package ms.konovalov.intellij.hidpi;

import com.intellij.ide.ui.UISettingsListener;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.*;
import com.intellij.openapi.editor.colors.EditorColorsManager;
import com.intellij.openapi.options.BaseConfigurableWithChangeSupport;
import com.intellij.util.messages.MessageBus;
import com.intellij.util.xmlb.XmlSerializerUtil;
import lombok.Getter;
import lombok.Setter;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

@State(name = "HIDPIProfiles", defaultStateAsResource = true, storages = @Storage(value = "hidpi.xml", roamingType = RoamingType.DISABLED))
public class FontSizeComponent extends BaseConfigurableWithChangeSupport implements ApplicationComponent, PersistentStateComponent<FontSizeComponent.State> {

    private FontSizeForm form;
    private State state = new State();

    public void initComponent() {
        if (state.getProfiles().isEmpty()) {
            FontProfile initialProfile = FontProfileManager.INSTANCE.readCurrentProfile("Default", true);
            state.getProfiles().add(initialProfile);
        }
        state.getActiveProfile().ifPresent(FontProfileManager.INSTANCE::applyProfile);

        MessageBus messageBus = ApplicationManager.getApplication().getMessageBus();
        messageBus.connect().subscribe(UISettingsListener.TOPIC, settingsChange ->
                state.getActiveProfile().ifPresent(active -> {
                    if (FontProfileManager.INSTANCE.changed(active, settingsChange)) {
                        FontProfileManager.INSTANCE.deselectAll();
                    }
                }));
        messageBus.connect().subscribe(EditorColorsManager.TOPIC, settingsChange ->
                state.getActiveProfile().ifPresent(active -> {
                    if (FontProfileManager.INSTANCE.changed(active, EditorColorsManager.getInstance().getGlobalScheme())) {
                        FontProfileManager.INSTANCE.deselectAll();
                    }
                }));
    }

    private List<FontProfile> profiles() {
        return state.getProfiles();
    }

    @Override
    public void disposeComponent() {
        if (state != null) {
            state.clear();
        }
    }

    @Override
    @NotNull
    public String getComponentName() {
        return "FontSizeComponent";
    }

    @Nullable
    @Override
    public State getState() {
        return state;
    }

    @Override
    public void loadState(State state) {
        XmlSerializerUtil.copyBean(state, this.state);
    }

    static List<FontProfile> getProfiles() {
        return ApplicationManager.getApplication().getComponent(FontSizeComponent.class).profiles();
    }

    @Getter
    @Setter
    static class State {

        private List<FontProfile> profiles = new ArrayList<>();

        void clear() {
            if (profiles != null) {
                profiles.clear();
            }
        }

        Optional<FontProfile> getActiveProfile() {
            return profiles.stream().filter(FontProfile::isActive).findFirst();
        }
    }

    @Nls
    @Override
    public String getDisplayName() {
        return "HIDPI Profiles";
    }

    @Nullable
    @Override
    public String getHelpTopic() {
        return "HIDPI Help";
    }

    @Nullable
    @Override
    public JComponent createComponent() {
        form = new FontSizeForm(this);
        return form.getPanel();
    }

    @Override
    public void apply() {
        form.apply();
    }

    @Override
    public void reset() {
        form.reset();
    }

    @Override
    public void disposeUIResources() {
        //do nothing
    }
}

File: src/main/java/ms/konovalov/intellij/hidpi/FontSizeComponent.kt
package ms.konovalov.intellij.hidpi

import com.intellij.ide.ui.UISettingsListener
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.*
import com.intellij.openapi.editor.colors.EditorColorsListener
import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.options.BaseConfigurableWithChangeSupport
import org.jetbrains.annotations.Nls
import java.util.*
import javax.swing.JComponent

@State(
    name = "HIDPIProfiles",
    defaultStateAsResource = true,
    storages = [(Storage(value = "hidpi.xml", roamingType = RoamingType.DISABLED))]
)
class FontSizeComponent : BaseConfigurableWithChangeSupport(), ApplicationComponent, PersistentStateComponent<FontSizeComponent.State> {

    private var form: FontSizeForm? = null
    private var state = State()

    override fun initComponent() {
        if (state.profiles.isEmpty()) {
            val initialProfile = FontProfileManager.readCurrentProfile("Default", true)
            state.profiles.add(initialProfile)
        }
        val active = state.activeProfile
        if (active != null) {
            FontProfileManager.applyProfile(active)
        }

        val messageBus = ApplicationManager.getApplication().messageBus
        messageBus.connect().subscribe(UISettingsListener.TOPIC, UISettingsListener { settingsChange ->
            if (active != null) {
                if (FontProfileManager.changed(active, settingsChange)) {
                    FontProfileManager.deselectAll()
                }
            }
        })
        messageBus.connect().subscribe(EditorColorsManager.TOPIC, EditorColorsListener { _ ->
            if (active != null) {
                if (FontProfileManager.changed(active, EditorColorsManager.getInstance().globalScheme)) {
                    FontProfileManager.deselectAll()
                }
            }
        })
    }

    override fun disposeComponent() {
        state.clear()
    }

    override fun getComponentName(): String {
        return "FontSizeComponent"
    }

    override fun getState(): State {
        return state
    }

    override fun loadState(state: State) {
        this.state = state
    }

    class State {
        var profiles: MutableList<FontProfile> = ArrayList()
        val activeProfile: FontProfile? = profiles.firstOrNull { it.isActive }

        fun clear() {
            this.profiles.clear()
        }
    }

    @Nls
    override fun getDisplayName(): String {
        return "HIDPI Profiles"
    }

    override fun getHelpTopic(): String? {
        return "HIDPI Help"
    }

    override fun createComponent(): JComponent? {
        form = FontSizeForm(this)
        return form!!.panel
    }

    override fun apply() {
        form!!.apply()
    }

    override fun reset() {
        form!!.reset()
    }

    override fun disposeUIResources() {
        //do nothing
    }
}
